src/audio_core/algorithm/interpolate.cpp

```diff
@@ -30,40 +148,42 @@ std::vector<s16> Interpolate(InterpolationState& state, std::vector<s16> input,
         ratio = 1.0;
     }
 
-    if (ratio != state.current_ratio) {
-        const double cutoff_frequency = std::min(0.5 / ratio, 0.5 * ratio);
-        state.nyquist = CascadingFilter::LowPass(std::clamp(cutoff_frequency, 0.0, 0.4), 3);
-        state.current_ratio = ratio;
-    }
-    state.nyquist.Process(input);
-
-    constexpr std::size_t taps = InterpolationState::lanczos_taps;
-    const std::size_t num_frames = input.size() / 2;
-
-    std::vector<s16> output;
-    output.reserve(static_cast<std::size_t>(input.size() / ratio + 4));
-
-    double& pos = state.position;
-    auto& h = state.history;
-    for (std::size_t i = 0; i < num_frames; ++i) {
-        std::rotate(h.begin(), h.end() - 1, h.end());
-        h[0][0] = input[i * 2 + 0];
-        h[0][1] = input[i * 2 + 1];
-
-        while (pos <= 1.0) {
-            double l = 0.0;
-            double r = 0.0;
-            for (std::size_t j = 0; j < h.size(); j++) {
-                const double lanczos_calc = Lanczos(taps, pos + j - taps + 1);
-                l += lanczos_calc * h[j][0];
-                r += lanczos_calc * h[j][1];
-            }
-            output.emplace_back(static_cast<s16>(std::clamp(l, -32768.0, 32767.0)));
-            output.emplace_back(static_cast<s16>(std::clamp(r, -32768.0, 32767.0)));
-
-            pos += ratio;
+    std::vector<s16> output(input.size() / ratio);
+
+    const int step = static_cast<int>(ratio * 0x8000);
+    const std::array<s16, 512>& lut = [ratio, step] {
+        if (step > 0xaaaa) {
+            return curve_lut0;
+        }
+        if (step <= 0x8000) {
+            return curve_lut1;
         }
-        pos -= 1.0;
+        return curve_lut2;
+    }();
+
+    int in_offset = 0;
+    for (int out_offset = 0; out_offset < output.size(); out_offset += 2) {
```

Comment:
```suggestion
    for (std::size_t out_offset = 0; out_offset < output.size(); out_offset += 2) {
```

Otherwise this will cause signedness mismatch warnings.